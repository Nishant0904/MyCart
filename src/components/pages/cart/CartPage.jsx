import styled from 'styled-components'
import Button from '../../reusables/buttons/Button'
import lockIcon from '../../../assets/icons/lock.svg'
import PropTypes from 'prop-types'
import CartProductItem from './CartProductItem'
import formatToUSD from '../../../utils/formatToUSD'
import groupProductsWithQuantities from '../../../utils/groupProductsWithQuantities'
import getSubtotal from '../../../utils/getSubtotal'

const CartPage = ({ cart, handleDeleteFromCart }) => {
    const cleanCart = groupProductsWithQuantities(cart)
    const subtotal = getSubtotal(cart)
    const freeShippingMin = 100
    const shipping = subtotal >= freeShippingMin ? 0 : 10

    if (cart.length === 0)
        return (
            <EmptyCartPage data-testid='empty-cart-page'>
                <h2>Your Cart is Empty</h2>
                Add products to your cart to see them appear here!
            </EmptyCartPage>
        )

    return (
        <StyledCartPage data-testid='cart-page'>
            <CartPageColumn>
                <h2>Your Cart</h2>
                <CartContainer>
                    {cleanCart.map((product, index) => {
                        return (
                            <CartProductItem
                                key={index}
                                product={product}
                                handleDeleteFromCart={handleDeleteFromCart}
                            />
                        )
                    })}
                </CartContainer>
            </CartPageColumn>
            <CartPageColumn>
                <h2>Summary</h2>
                <p>
                    {subtotal >= freeShippingMin
                        ? 'Woohoo! You qualify for free shipping!'
                        : `Spend ${formatToUSD(freeShippingMin - subtotal)} more to qualify for free shipping.`}
                </p>
                <PriceSummaryContainer>
                    <div>
                        <p>Subtotal</p>
                        <p>{formatToUSD(subtotal)}</p>
                    </div>
                    <div>
                        <p>
                            {subtotal >= freeShippingMin
                                ? 'Free Shipping'
                                : 'Flat Shipping Rate'}
                        </p>
                        <p>{formatToUSD(shipping)}</p>
                    </div>
                    <div>
                        <p>Total</p>
                        <p>{formatToUSD(subtotal + shipping)}</p>
                    </div>
                </PriceSummaryContainer>
                <Button>
                    <img src={lockIcon} alt='lock icon' color='white' />
                    Checkout Securely
                </Button>
            </CartPageColumn>
        </StyledCartPage>
    )
}

const EmptyCartPage = styled.section`
    flex-grow: 1;
    display: flex;
    flex-direction: column;
    gap: 8px;
    align-items: center;
    justify-content: center;
`

const StyledCartPage = styled.section`
    display: grid;
    grid-template-columns: 2fr 1fr;
    gap: 48px;
    padding: 48px 96px;

    @media (max-width: 740px) {
        padding: 48px 24px;
        grid-template-columns: 1fr;
    }
`

const CartPageColumn = styled.div`
    display: flex;
    flex-direction: column;
    gap: 24px;
    @media (max-width: 740px) {
        &:first-child {
            order: 2;
        }
    }
`

const CartContainer = styled.ul`
    display: flex;
    flex-direction: column;
    list-style: none;

    li + li {
        border-top: 1px solid grey;
    }
`

const PriceSummaryContainer = styled.div`
    display: flex;
    flex-direction: column;
    gap: 8px;

    div {
        display: flex;
        justify-content: space-between;
    }
`

CartPage.propTypes = {
    cart: PropTypes.array.isRequired,
    handleDeleteFromCart: PropTypes.func.isRequired,
}

export default CartPage
