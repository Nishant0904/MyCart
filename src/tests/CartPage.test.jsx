/* eslint-disable react/prop-types */
import { describe, it, expect, vi } from 'vitest'
import { render, screen } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import CartPage from '../components/pages/cart/CartPage'

const mockCart = [
    {
        id: 1,
        title: 'Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops',
        price: 109.95,
        description:
            'Your perfect pack for everyday use and walks in the forest. Stash your laptop (up to 15 inches) in the padded sleeve, your everyday',
        category: "men's clothing",
        image: 'https://fakestoreapi.com/img/81fPKd-2AYL._AC_SL1500_.jpg',
        rating: {
            rate: 3.9,
            count: 120,
        },
    },
]

vi.mock('../components/pages/cart/CartProductItem.jsx', () => ({
    default: ({ product, handleDeleteFromCart }) => {
        return (
            <div data-testid='cart-product-item-mock'>
                <h4>{product.title}</h4>
                <button onClick={handleDeleteFromCart}></button>
            </div>
        )
    },
}))

describe('Cart Page test', () => {
    it('renders correctly with an empty cart', () => {
        render(<CartPage cart={[]} handleDeleteFromCart={() => {}} />, {
            wrapper: BrowserRouter,
        })
        expect(screen.getByTestId('empty-cart-page')).toMatchSnapshot()
        expect(
            screen.getByRole('heading', {
                level: 2,
                name: 'Your Cart is Empty',
            }),
        ).toBeInTheDocument()
    })

    it('renders correctly with a full cart', () => {
        render(<CartPage cart={mockCart} handleDeleteFromCart={() => {}} />, {
            wrapper: BrowserRouter,
        })
        expect(screen.getByTestId('cart-page')).toMatchSnapshot()
        expect(
            screen.getByRole('heading', {
                level: 2,
                name: 'Your Cart',
            }),
        ).toBeInTheDocument()
        expect(
            screen.getByRole('heading', {
                level: 2,
                name: 'Summary',
            }),
        ).toBeInTheDocument()
    })

    it('renders Cart Product Item', () => {
        render(<CartPage cart={mockCart} handleDeleteFromCart={() => {}} />, {
            wrapper: BrowserRouter,
        })

        expect(screen.getByTestId('cart-product-item-mock')).toBeInTheDocument()
    })
})
