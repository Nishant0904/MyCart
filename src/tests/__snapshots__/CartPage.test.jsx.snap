// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Cart Page test > renders correctly with a full cart 1`] = `
<section
  class="sc-dmyCSP jeNyzD"
  data-testid="cart-page"
>
  <div
    class="sc-hLQSwg fKaRla"
  >
    <h2>
      Your Cart
    </h2>
    <ul
      class="sc-eDLKkx buiyLQ"
    >
      <div
        data-testid="cart-product-item-mock"
      >
        <h4>
          Fjallraven - Foldsack No. 1 Backpack, Fits 15 Laptops
        </h4>
        <button />
      </div>
    </ul>
  </div>
  <div
    class="sc-hLQSwg fKaRla"
  >
    <h2>
      Summary
    </h2>
    <p>
      Woohoo! You qualify for free shipping!
    </p>
    <div
      class="sc-jTQCzO ieSRxs"
    >
      <div>
        <p>
          Subtotal
        </p>
        <p>
          $109.95
        </p>
      </div>
      <div>
        <p>
          Free Shipping
        </p>
        <p>
          $0.00
        </p>
      </div>
      <div>
        <p>
          Total
        </p>
        <p>
          $109.95
        </p>
      </div>
    </div>
    <button
      class="sc-beySPh gepDOA"
    >
      <img
        alt="lock icon"
        color="white"
        src="/src/assets/icons/lock.svg"
      />
      Checkout Securely
    </button>
  </div>
</section>
`;

exports[`Cart Page test > renders correctly with an empty cart 1`] = `
<section
  class="sc-guDLey eVLQBN"
  data-testid="empty-cart-page"
>
  <h2>
    Your Cart is Empty
  </h2>
  Add products to your cart to see them appear here!
</section>
`;
